src/models/stores/documents.ts

```diff
@@ -174,6 +184,33 @@ export const DocumentsModel = types
       }
     };
 
+    const addRequiredDocumentPromises = (requiredTypes: string[]) => {
+      requiredTypes.forEach(type => {
+        const wrapper: Partial<IRequiredDocumentPromise> = { isResolved: false };
+        wrapper.promise = new Promise(resolve => {
+          wrapper.resolve = (document: DocumentModelType | null) => {
+            resolve(document);
+            wrapper.isResolved = true;
+          };
+        });
```

Comment:
This is tricky to understand. What I understand is that this is basically exposing a promises internal resolve method so it can be called by something outside of the promise. And the isResolved property monitors this state. 

I did some searching and found this pattern is called `Deferred` in some cases: https://stackoverflow.com/a/34637436